src/components/chat/IncomingResponseMessage.tsx

```diff
@@ -27,30 +27,87 @@ export const IncomingResponseMessage = ({
     >
       <div>
         <p>{message.textBeforeData}</p>
-        {message.data && message.data.length && (
-          <div className='m-2 mt-4 max-h-[300px] overflow-auto'>
-            <Table striped bordered hover>
-              <thead>
-                <tr>
-                  {Object.keys(message.data[0]).map(key => (
-                    <th key={key}>{key}</th>
-                  ))}
-                </tr>
-              </thead>
-              <tbody>
-                {message.data.map((row, index) => (
-                  <tr key={index}>
-                    {Object.keys(message.data![0]).map(key => (
-                      <td key={key}>{row[key]}</td>
-                    ))}
-                  </tr>
-                ))}
-              </tbody>
-            </Table>
-          </div>
-        )}
+        {message.data?.length && <ChatDataTable data={message.data} />}
         {message.textAfterData && <p>{message.textAfterData}</p>}
       </div>
     </IncomingMessageWithFeedbackButtons>
   );
 };
+
+type ChatDataTableProps = {
+  data: { [key: string]: string | number }[];
+};
+
+const ChatDataTable = ({ data }: ChatDataTableProps) => {
+  const [paginationModel, setPaginationModel] = React.useState({
+    pageSize: 5,
+    page: 0,
+  });
+
+  function deriveHeaders({ data }: ChatDataTableProps) {
+    const headerFormats = [
+      {
+        key: 'proportion',
+        name: 'Proportion (%)',
+        valueFormatter: (params: any) => `${(params.value * 100).toFixed(2)}`,
+      },
+    ];
+
+    const headerData = Object.keys(data[0]);
+    return headerData.map(key => {
+      return {
+        field: key,
+        flex: 1,
+        headerName: headerFormats.find(headerName => headerName.key === key)?.name || key,
+        valueFormatter: headerFormats.find(headerName => headerName.key === key)?.valueFormatter,
+      };
+    });
+  }
+
+  function deriveRows({ data }: ChatDataTableProps) {
+    return data.map((row, index) => {
+      return { id: index, ...row };
+    });
+  }
+
+  const headers = deriveHeaders({ data });
+  const rows = deriveRows({ data });
+
+  function getHeightOfTable(numberOfRows: number = rows.length) {
+    const heightFooter = 3.5;
+    const heightHeader = heightFooter;
+    const heightToolbar = heightHeader;
+    const heightLine = 3.25;
+    return (
+      Math.min(numberOfRows, paginationModel.pageSize) * heightLine +
+      heightFooter +
+      heightHeader +
+      heightToolbar
+    );
+  }
+
+  return (
+    <div style={{ height: `${getHeightOfTable()}rem`, backgroundColor: 'background.paper' }}>
+      <DataGrid
+        columns={headers}
+        rows={rows}
+        autoHeight={true}
+        paginationModel={paginationModel}
+        pageSizeOptions={[5, 10, 20]}
+        onPaginationModelChange={model => setPaginationModel(model)}
+        sx={{
+          backgroundColor: 'background.paper',
+        }}
+        slots={{ toolbar: CustomToolbar }}
```

Comment:
I think that it would be better to save some space and make the rows as narrow as possible